### PR TITLE
Make it explicit that contract key maintainers are computed from key

### PR DESCRIPTION
**Pull request by hurryabit**
Until now, you could write things like
```
  key (bank, accountId) : (Party, Text)
  maintainer bank
```
Since the maintainer needs to be computed from the `key` value in DAML-LF, this required some magical rewriting of the `maintainer` expression in the compiler. This rewriting is brittle, particularly in the presence of complex `key` expressions, and lead to horrible error messages.

This PR makes the fact that the `maintainer` is computed from the `key` explicit. Now, you have to write
```
  key (bank, accountId) : (Party, Text)
  maintainer key._1
```
to get the same behaviour as above.

Removing the underlying magic from the compiler (and the fact the contract keys can be arbitrary expressions now) allows for removing a few tests that are no long meaningful.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/1527)

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
I like this a lot. Can the user make the contract key a record so that it can refer to fields by name rather then by position?

File: daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/Template.daml [hunk 1]
```diff
@@ -184,7 +184,7 @@ class Template c => TemplateKey c k | c -> k where
     key : c -> k
 
     -- | The maintainers of the contract key.
-    maintainer : c -> [Party]
+    maintainer : k -> [Party]
```
Comment (shayne-fletcher-da) [inline daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/Template.daml]: Can the documentation in `//ghc-lib/template-desugaring.md` get an update please or if you prefer, you can instruct me to do it.

**Comment by hurryabit** [comment]
@stefanobaghino-da Yes, the key expression can be of any type, e.g., `data AccountKey = AccountKey with bank : Party; accountId : Text`. Then you could have `maintainer key.bank`. I think that is what @bame-da wants to do.

File: daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/Template.daml (hunk 1 above)
Comment (hurryabit) [reply daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/Template.daml]: I updated it.

**Pull request merged**